PR implement OTS calculate checksum

pirun opened the pull request:
Implement OTS OACP Calculate Checksum Procedure.
Add checksum usage on both central and peripheral sample.

Thalley commented on include/zephyr/bluetooth/services/ots.h:
```diff
@@ -725,6 +725,19 @@ struct bt_ots_cb {
+	 *
+	 *  @param id         Object ID.
+	 *  @param offset     The first octet of the object contents need to be calculated.
+	 *  @param length     The length number of octets object name.
+	 */
+	uint32_t (*obj_cal_checksum)(uint64_t id, off_t offset, size_t len);
```
Missing documentation of the return value. I assume it should return the checksum? 

Since the callback contains offset and len, does the functionality work such that a client can request a checksum of a subset of the object?

pirun replied:
1. Description of return will be added.

1. Yes, Calculate Checksum Procedure could request a subset of the object.

![image](https://user-images.githubusercontent.com/1169294/192174220-6d9a96b5-6853-4208-a688-fb531bf17344.png)

Thalley commented on include/zephyr/bluetooth/services/ots.h:
```diff
@@ -888,6 +901,20 @@ struct bt_ots_client_cb {
+	 *  @param err               Error code (bt_ots_olcp_res_code).
+	 *  @param crc               Checksum code (bt_ots_olcp_res_code).
+	 */
+	void (*obj_checksum_got)(struct bt_ots_client *ots_inst,
```
```suggestion
	void (*obj_checksum)(struct bt_ots_client *ots_inst,
```

pirun replied:
We use past principle naming of OTS client callbacks.

> e.g.:`obj_selected`,`obj_data_written`.

May be change it to `obj_checksum_calculated`?

Thalley replied:
That could also work :)

pirun replied:
Change to ``obj_checksum_calculated``.

Thalley commented on include/zephyr/bluetooth/services/ots.h:
```diff
@@ -888,6 +901,20 @@ struct bt_ots_client_cb {
+	 *  @param ots_inst          Pointer to the OTC instance.
+	 *  @param conn              The connection to the peer device.
+	 *  @param err               Error code (bt_ots_olcp_res_code).
+	 *  @param crc               Checksum code (bt_ots_olcp_res_code).
+	 */
+	void (*obj_checksum_got)(struct bt_ots_client *ots_inst,
+				 struct bt_conn *conn, int err, uint32_t crc);
```
Is `crc` the best name for the value? Wouldn't it make more sense to just use `checksum`?

pirun replied:
Will revise.

pirun replied:
Change to ``checksum``.

Thalley commented on include/zephyr/bluetooth/services/ots.h:
```diff
@@ -1029,6 +1056,23 @@ int bt_ots_client_write_object_data(struct bt_ots_client *otc_inst, struct bt_co
+ *
+ *  @param otc_inst     Pointer to the OTC instance.
+ *  @param conn         Pointer to the connection object.
+ *  @param offset       Offset to calculate, usually 0.
+ *  @param len          Len of data.
```
Is there any way that the client can request the checksum without knowing the length of the object?

pirun replied:
No, I think it is a must.
Quote from OTS SPEC 

> A Length parameter is required with the Calculate Checksum Op Code. The object contents included in the calculation shall consist of Length number of octets, starting with the octet identified by the value of the Offset parameter.

Length can be obtained by Object Size characteristic.

Thalley replied:
I see.

Given that we can do a checksum on a subset of the object, I guess `len` just needs to be `<= object_size`.

Thalley commented on samples/bluetooth/central_otc/src/main.c:
```diff
@@ -568,6 +588,10 @@ static int on_obj_data_read(struct bt_ots_client *ots_inst, struct bt_conn *conn
 		printk("Object total received %d\n", len + offset);
 		print_hex_number(obj_data_buf, len + offset);
 		(void)memset(obj_data_buf, 0, OBJ_MAX_SIZE);
+		otc_crc_work.offset = 0;
+		otc_crc_work.len = otc.cur_object.size.cur;
+		k_work_schedule(&otc_crc_work.work, K_MSEC(50));
```
What if `&otc_crc_work.work` is already being used to calculate another checksum?

pirun replied:
oacp_checksum should return -EBUSY if there is already called.

Thalley commented on samples/bluetooth/central_otc/src/main.c:
```diff
@@ -151,17 +157,31 @@ static void otc_btn_work_fn(struct k_work *work)
 	}
 }
 
+static void otc_crc_work_fn(struct k_work *work)
```
Ditto w.r.t. using `crc` vs `checksum`; suggest to use `checksum`

pirun replied:
Change to ``checksum``.

Thalley commented on samples/bluetooth/central_otc/src/main.c:
```diff
@@ -592,12 +616,19 @@ static void on_obj_data_written(struct bt_ots_client *ots_inst, struct bt_conn *
+void on_obj_checksum_got(struct bt_ots_client *ots_inst,
+			 struct bt_conn *conn, int err, uint32_t crc)
+{
+	printk("Object checksum OACP result (%d) crc 0x%08x\n", err, crc);
```
Shouldn't we do our own local checksum to compare here? A checksum without a comparison is rather useless :)

pirun replied:
Yes, I will add comparison local and remote checksum here.

pirun replied:
Done.

Thalley commented on samples/bluetooth/peripheral_ots/src/main.c:
```diff
@@ -192,13 +193,24 @@ static void ots_obj_name_written(struct bt_ots *ots, struct bt_conn *conn,
 		id_str, cur_name, new_name);
 }
 
+uint32_t ots_obj_cal_checksum(uint64_t id, off_t offset, size_t len)
+{
+	uint32_t crc;
+	uint32_t obj_index = ((id - BT_OTS_OBJ_ID_MIN) % ARRAY_SIZE(objects));
+
+	crc = crc32_ieee((const uint8_t *)(&objects[obj_index].data + offset), len);
+	printk("offset %lu len %zu crc 0x%08x\n", offset, len, crc);
+	return crc;
+}
```
I assume this is calculating the checksum with accordance to the OTS spec. 

Suggest to add a new function, ala `bt_ots_cal_checksum` that can be used so that the application doesn't need to implement it itself

pirun replied:
I can't make my mind about where to implement the **"actual"** checksum calculation.

1. Let application calculate.
2. Application return address of data, let OTS service to calculate.

Since object data is opaque, OTS service can't access object data directly.

Thalley replied:
Yeah, so basically we have

1: The application implements and calculates the checksum. However, since it needs to be return in this callback, we will be blocking whatever thread need that needs to return this value. IIRC an object can be up to 2^32-1 octets in size (roughly 4GB), and calculating the checksum of such an object could potentially take time, so calculating it in a callback isn't perhaps the best solution.

2: Application returns the pointer to the object's data, and we can implement the calculation and handling of calculating this in a different thread than the BT RX thread (e.g. the BT Long Workqueue), and send the indication when done

3: Like 1:, but where we expose a `bt_ots_calculate_checksum` function to the application, so that it doesn't need to worry about using `crc32_ieee` correctly. 

I think 2: would give the best result, as we can best control the calculation there, and it follows the same principles of the read functionality.

Thalley commented on include/zephyr/bluetooth/services/ots.h:
```diff
@@ -725,6 +725,20 @@ struct bt_ots_cb {
+	 *  @param length     The length number of octets object name.
+	 *  @return Pointer of Object data.
+	 */
+	const uint8_t *(*obj_cal_checksum)(uint64_t id, off_t offset, size_t len);
```
`obj_read` and `obj_write` use `void *` for the data pointers, suggest to do the same here. 

We could also consider changing the return value to `ssize_t` or `int` so that the application can return an error if something isn't correct, or it wants to reject the request for whatever reason (e.g. if the `len` is too large). 

In other words, I suggest that this works the same as the `obj_read`, including the `ots` and `conn` pointers too.

pirun replied:
Modify OTS client API to int bt_ots_client_get_object_checksum(struct bt_ots_client *otc_inst, struct bt_conn *conn,
				      off_t offset, size_t len);

Thalley commented on include/zephyr/bluetooth/services/ots.h:
```diff
@@ -888,6 +902,19 @@ struct bt_ots_client_cb {
+	 *  @param ots_inst          Pointer to the OTC instance.
+	 *  @param conn              The connection to the peer device.
+	 *  @param err               Error code (bt_ots_olcp_res_code).
+	 *  @param checksum          Checksum code (bt_ots_olcp_res_code).
```
I assume that `checksum` will be `0` if `err != 0`?

pirun replied:
Modify description to 
`	 *  @param checksum          Checksum if error code is BT_GATT_OTS_OACP_RES_SUCCESS,
	 *                           otherwise 0.`

Thalley commented on include/zephyr/bluetooth/services/ots.h:
```diff
@@ -1029,6 +1056,23 @@ int bt_ots_client_write_object_data(struct bt_ots_client *otc_inst, struct bt_co
+ *  This will trigger an OACP calculate checksum operation for the current object
+ *  with a specified offset and length.
+ *
+ *  The checksum goes to OACP IND and obj_data_checksum() callback.
```
Does the `obj_data_checksum()` reference work the callback? Don't you need to do something like `bt_ots_client_cb.obj_data_checksum()` or similar to generate a proper reference to the callback in the doxygen documentation?

pirun replied:
Modify description to
` *  The checksum goes to OACP IND and obj_checksum_calculated() callback.`

Thalley commented on include/zephyr/bluetooth/services/ots.h:
```diff
@@ -1029,6 +1056,23 @@ int bt_ots_client_write_object_data(struct bt_ots_client *otc_inst, struct bt_co
+ *  @param otc_inst     Pointer to the OTC instance.
+ *  @param conn         Pointer to the connection object.
+ *  @param offset       Offset to calculate, usually 0.
+ *  @param len          Len of data.
```
```suggestion
 *  @param len          Len of data to calculate checksum for. May be less than the current object's size, but shall not be larger. 
```

pirun replied:
Done.

Thalley commented on samples/bluetooth/central_otc/src/main.c:
```diff
@@ -134,12 +140,13 @@ static void otc_btn_work_fn(struct k_work *work)
 				obj_data_buf[idx] = UINT8_MAX - (idx % UINT8_MAX);
 			}
 
+			last_checksum = crc32_ieee(obj_data_buf, size_to_write);
```
I would suggest to move the `crc32_ieee` to a `bt_ots_client_calc_checksum` function so that the client application do not need to worry about how to calculate it correctly.

Thalley replied:
I somewhat agree, but my point is that with this approach, the application needs to know exactly which algorithm to use to calculate the checksum (in this case, the one implemented by `crc32_ieee`, but if there was a `bt_ots_client_calc_checksum` function, that knowledge could (easily) be abstracted, making it easier to use the OTC. 

If you disagree, then we should at least (if not already) have a reference to `crc32_ieee` in the documentation.

pirun replied:
Add bt_ots_client_calc_checksum function as a wrapper to call crc32_ieee().

Thalley commented on samples/bluetooth/central_otc/src/main.c:
```diff
@@ -568,6 +590,10 @@ static int on_obj_data_read(struct bt_ots_client *ots_inst, struct bt_conn *conn
 		printk("Object total received %d\n", len + offset);
 		print_hex_number(obj_data_buf, len + offset);
 		(void)memset(obj_data_buf, 0, OBJ_MAX_SIZE);
+		otc_checksum_work.offset = 0;
+		otc_checksum_work.len = otc.cur_object.size.cur;
+		k_work_schedule(&otc_checksum_work.work, K_MSEC(50));
+
```
Why do we need a workqueue item for this? At this point we are just sending a GATT request in the `otc_checksum_work_fn` function, and that should be fine to do in this callback directly, right?

pirun replied:
OTS client API couldn't be called in OTS client cb function.
obj_data_read cb is called [here](https://github.com/zephyrproject-rtos/zephyr/blob/main/subsys/bluetooth/services/ots/ots_client.c#L194) before [cur_inst released](https://github.com/zephyrproject-rtos/zephyr/blob/main/subsys/bluetooth/services/ots/ots_client.c#L209).
Which makes oacp_checksum return -EBUSY [here](https://github.com/zephyrproject-rtos/zephyr/pull/50451/files#diff-7278e3a1c0d69370760b526260965a6fd17f04254467a377f08592d1c11ccb6aR1302) .

Thalley replied:
Fair point. Is that mentioned in the API documentation anywhere?

pirun replied:
It doesn't in API documentation.

Thalley commented on subsys/bluetooth/services/ots/ots_client.c:
```diff
@@ -1274,6 +1286,50 @@ static int oacp_write(struct bt_conn *conn, struct bt_otc_internal_instance_t *i
+	inst->otc_inst->write_params.func = write_oacp_cp_cb;
+
+	err = bt_gatt_write(conn, &inst->otc_inst->write_params);
+
+	if (err != 0) {
```
```suggestion
	if (err != 0) {
```

pirun replied:
Done.

Thalley commented on subsys/bluetooth/services/ots/ots_client.c:
```diff
@@ -1370,6 +1426,55 @@ int bt_ots_client_write_object_data(struct bt_ots_client *otc_inst,
+		return -EINVAL;
+	}
+
+	inst = lookup_inst_by_handle(otc_inst->start_handle);
+
+	if (!inst) {
```
```suggestion
	inst = lookup_inst_by_handle(otc_inst->start_handle);
	if (!inst) {
```

pirun replied:
Done.

Thalley commented on subsys/bluetooth/services/ots/ots_oacp.c:
```diff
@@ -23,7 +25,7 @@
 
 #define OACP_PROC_TYPE_SIZE	1
-#define OACP_RES_MAX_SIZE	3
+#define OACP_RES_MAX_SIZE	7
```
Increased by 4 to support the 4-octet checksum?

pirun replied:
done

Thalley commented on subsys/bluetooth/services/ots/ots_oacp.c:
```diff
@@ -161,6 +163,51 @@ static enum bt_gatt_ots_oacp_res_code oacp_delete_proc_validate(
+	LOG_DBG("Validating Checksum procedure with offset: 0x%08X and "
+		"length: 0x%08X", params->offset, params->len);
+
+	CHECKIF(!ots->cur_obj) {
+		return BT_GATT_OTS_OACP_RES_INV_OBJ;
+	}
+
+	CHECKIF(params->offset > ots->cur_obj->metadata.size.cur) {
+		return BT_GATT_OTS_OACP_RES_INV_PARAM;
+	}
+
+	CHECKIF((params->offset + (uint64_t) params->len) > ots->cur_obj->metadata.size.alloc) {
+		return BT_GATT_OTS_OACP_RES_INV_PARAM;
+	}
+
+	CHECKIF(ots->cur_obj->state.type != BT_GATT_OTS_OBJECT_IDLE_STATE) {
+		return BT_GATT_OTS_OACP_RES_OBJ_LOCKED;
+	}
```
`CHECKIF` should not be used to verify data from remote devices, as it can be configured to work as an ASSERT, and we do not want to ASSERT if a remote device sends bad data. 

Suggest to just use `if` here

pirun replied:
Done.

Thalley commented:
@pirun I'll be busy preparing for and attending UPF the next two weeks, so my next review will be sometime after that. I'll dismiss my change requests in case someone else approves this :)

Thalley commented on samples/bluetooth/peripheral_ots/src/main.c:
```diff
@@ -192,13 +192,27 @@ static void ots_obj_name_written(struct bt_ots *ots, struct bt_conn *conn,
+static int ots_obj_cal_checksum(struct bt_ots *ots, struct bt_conn *conn, uint64_t id,
+				void **data, off_t offset, size_t len)
+{
+	uint32_t obj_index = ((id - BT_OTS_OBJ_ID_MIN) % ARRAY_SIZE(objects));
```
And I also see that you have
```c
	if (obj_index > OBJ_POOL_SIZE) {
		return -ENOENT;
	}
``` 
right below, so the modulo doesn't seem to have any purpose.

Thalley commented on subsys/bluetooth/services/ots/ots_client.c:
```diff
@@ -1359,6 +1414,54 @@ int bt_ots_client_write_object_data(struct bt_ots_client *otc_inst,
+	}
+
+	CHECKIF((offset > UINT32_MAX) || (offset < 0)) {
+		LOG_ERR("offset exceeds UINT32");
```
```suggestion
		LOG_ERR("offset exceeds UINT32 and must be >= 0");
```

pirun replied:
done

Thalley commented on subsys/bluetooth/services/ots/ots_client.c:
```diff
@@ -1359,6 +1414,54 @@ int bt_ots_client_write_object_data(struct bt_ots_client *otc_inst,
+	 *	Offset and Length field are UINT32 Length
+	 */
+	CHECKIF(len > UINT32_MAX) {
+		LOG_ERR("length exceeds UINT32");
```
```suggestion
		LOG_ERR("length %zu exceeds UINT32", len);
```

Thalley commented on subsys/bluetooth/services/ots/ots_client.c:
```diff
@@ -1359,6 +1414,54 @@ int bt_ots_client_write_object_data(struct bt_ots_client *otc_inst,
+	}
+
+	CHECKIF((offset > UINT32_MAX) || (offset < 0)) {
+		LOG_ERR("offset exceeds UINT32");
```
```suggestion
		LOG_ERR("offset %d exceeds UINT32", offset);
```

pirun replied:
done

Thalley commented on subsys/bluetooth/services/ots/ots_client.c:
```diff
@@ -1359,6 +1414,54 @@ int bt_ots_client_write_object_data(struct bt_ots_client *otc_inst,
+	}
+
+	CHECKIF((len + offset) > otc_inst->cur_object.size.cur) {
+		LOG_ERR("The sum of Offset and Length exceed the Current Size %lu alloc %zu.",
+		(len + offset),	otc_inst->cur_object.size.cur);
```
```suggestion
		LOG_ERR("The sum of offset (%d) and length (%zu) exceed the Current Size %lu alloc %zu.",
		offset, len, (len + offset), otc_inst->cur_object.size.cur);
```

pirun replied:
done

Thalley commented on subsys/bluetooth/services/ots/ots_oacp.c:
```diff
@@ -23,7 +25,7 @@
 
 #define OACP_PROC_TYPE_SIZE	1
-#define OACP_RES_MAX_SIZE	3
+#define OACP_RES_MAX_SIZE	(3 + sizeof(uint32_t))
```
Please add a comment stating why it's 3 + 4 (i.e. what fields are making this MAX size)

pirun replied:
done

Thalley replied:
Need more details: Why is it 3?

pirun replied:
Add complete comment of OACP response.

Thalley commented on subsys/bluetooth/services/ots/ots_oacp.c:
```diff
@@ -161,6 +163,53 @@ static enum bt_gatt_ots_oacp_res_code oacp_delete_proc_validate(
 #endif
 
+#if defined(CONFIG_BT_OTS_OACP_CHECKSUM_SUPPORT)
+static enum bt_gatt_ots_oacp_res_code oacp_checksum_proc_validate(
```
This functions also calls the `obj_cal_checksum` callback, so it seems to do more than just validate. Please either rename, or remove the call to the callback.

Thalley replied:
Oh, I see that we use a similar naming for the other functions that also does more than just validate. IMO that's not very good, as `validate` doesn't not imply any state changes.

pirun replied:
Just try to follow the original code structure and naming rule.

Thalley replied:
Yeah, I saw that. It is OK as is for now, but IMO should be updated in the future.

pirun replied:
We need to refactor this one day.

Thalley commented on subsys/bluetooth/services/ots/ots_oacp.c:
```diff
@@ -161,6 +163,53 @@ static enum bt_gatt_ots_oacp_res_code oacp_delete_proc_validate(
+static enum bt_gatt_ots_oacp_res_code oacp_checksum_proc_validate(
+	struct bt_conn *conn,
+	struct bt_ots *ots,
+	struct bt_gatt_ots_oacp_proc *proc)
+{
+	struct bt_gatt_ots_oacp_cs_calc_params *params = &proc->cs_calc_params;
```
```suggestion
	const struct bt_gatt_ots_oacp_proc *proc)
{
	const struct bt_gatt_ots_oacp_cs_calc_params *params = &proc->cs_calc_params;
```

pirun replied:
Const qualifier can't be assigned since member checksum_resp of params will be changed in this function.

Thalley replied:
Why do we store the result in the parameters? That seems like a misuse of the word "parameter".

pirun replied:
You're right. I misused that parameter to store result.
Add function parameter to retrieve checksum sum.

pirun commented on include/zephyr/bluetooth/services/ots.h:
```diff
@@ -725,6 +725,24 @@ struct bt_ots_cb {
+	 *  @param ots      OTS instance.
+	 *  @param conn     The connection that wrote object.
+	 *  @param id       Object ID.
+	 *  @param data     Out: Pointer of actual object data.
```
Do we denote input/output parameter when it comes to callback function?

Thalley replied:
> Do we denote input/output parameter when it comes to callback function?

I assume so? Would make sense to use the same convention there, right?

pirun replied:
Add ``[in]`` and ``[out]`` to param for this function.

pirun commented on subsys/bluetooth/services/ots/ots_client.c:
```diff
@@ -1359,6 +1414,54 @@ int bt_ots_client_write_object_data(struct bt_ots_client *otc_inst,
+	 *	Offset and Length field are UINT32 Length
+	 */
+	CHECKIF(len > UINT32_MAX) {
+		LOG_ERR("length exceeds UINT32");
```
We have several length checks in different function, should all of these check change to LOG_DBG?

Thalley replied:
I would say so. LOG_ERR should only be used rarely as it would typically be enabled and take up significant space

pirun replied:
I can only change ``LOG_ERR`` to ``LOG_DBG`` for ``ots_obj_cal_checksum``, since the other functions are not in scope of this PR.

pirun commented on include/zephyr/bluetooth/services/ots.h:
```diff
@@ -725,6 +725,24 @@ struct bt_ots_cb {
+	/** @brief Object Calculate checksum callback
+	 *
+	 *  This callback is called when the OACP Calculate Checksum procedure is performed.
+	 *  Because object data is opaque to OTS service, the OTS user is the only one who
```
Done.

pirun commented on include/zephyr/bluetooth/services/ots.h:
```diff
@@ -725,6 +725,24 @@ struct bt_ots_cb {
+	/** @brief Object Calculate checksum callback
+	 *
+	 *  This callback is called when the OACP Calculate Checksum procedure is performed.
+	 *  Because object data is opaque to OTS service, the OTS user is the only one who
+	 *  knows where data is, user should return pointer of actual object data.
```
Done.

pirun commented on include/zephyr/bluetooth/services/ots.h:
```diff
@@ -725,6 +725,24 @@ struct bt_ots_cb {
+	 *  @param id       Object ID.
+	 *  @param data     Out: Pointer of actual object data.
+	 *  @param offset   The first octet of the object contents need to be calculated.
+	 *  @param len      The length number of octets object name.
+	 *  @return         Zero on success retrieve object data from OTS user.
+	 *  @return         A negative value in case of an error.
```
Done.

pirun commented on samples/bluetooth/central_otc/src/main.c:
```diff
@@ -568,6 +589,10 @@ static int on_obj_data_read(struct bt_ots_client *ots_inst, struct bt_conn *conn
 		(void)memset(obj_data_buf, 0, OBJ_MAX_SIZE);
+		otc_checksum_work.offset = 0;
+		otc_checksum_work.len = otc.cur_object.size.cur;
+		k_work_schedule(&otc_checksum_work.work, K_MSEC(50));
```
Done.

pirun commented on subsys/bluetooth/services/ots/ots_client.c:
```diff
@@ -369,6 +371,16 @@ static void oacp_ind_handler(struct bt_conn *conn,
 		enum bt_gatt_ots_oacp_res_code  result_code =
 			net_buf_simple_pull_u8(&net_buf);
+		if (req_opcode == BT_GATT_OTS_OACP_PROC_CHECKSUM_CALC) {
+			uint32_t checksum = net_buf_simple_pull_le32(&net_buf);
```
Done.

Thalley commented on include/zephyr/bluetooth/services/ots.h:
```diff
@@ -1082,6 +1133,23 @@ static inline int bt_ots_obj_id_to_str(uint64_t obj_id, char *str, size_t len)
+ * @brief  Generate IEEE conform CRC32 checksum.
+ *
+ * To abstract IEEE implementation to service layer.
+ * Application doesn't need to understand how crc32 is generated.
```
```suggestion
```
Doesn't seem useful to say

pirun replied:
done

Thalley commented on include/zephyr/bluetooth/services/ots.h:
```diff
@@ -1082,6 +1133,23 @@ static inline int bt_ots_obj_id_to_str(uint64_t obj_id, char *str, size_t len)
+ * To abstract IEEE implementation to service layer.
+ * Application doesn't need to understand how crc32 is generated.
+ *
+ * @param  *data        Pointer to data on which the CRC should be calculated.
```
```suggestion
 * @param  data        Pointer to data on which the CRC should be calculated.
```

pirun replied:
done

Thalley commented on samples/bluetooth/peripheral_ots/src/main.c:
```diff
@@ -43,6 +43,8 @@ struct object_creation_data {
 	uint32_t props;
 };
 
+#define OTS_OBJ_ID_TO_OBJ_IDX(id) ((id - BT_OTS_OBJ_ID_MIN) % ARRAY_SIZE(objects))
```
```suggestion
#define OTS_OBJ_ID_TO_OBJ_IDX(id) (((id) - BT_OTS_OBJ_ID_MIN) % ARRAY_SIZE(objects))
```

pirun replied:
done

Thalley commented on subsys/bluetooth/services/ots/ots_client.c:
```diff
@@ -365,10 +370,29 @@ static void oacp_ind_handler(struct bt_conn *conn,
+		}
+
+		if (req_opcode == BT_GATT_OTS_OACP_PROC_CHECKSUM_CALC) {
+			if (net_buf.len >= sizeof(uint32_t)) {
```
```suggestion
			if (net_buf.len == sizeof(checksum)) {
```

pirun replied:
done

Thalley commented on subsys/bluetooth/services/ots/ots_client.c:
```diff
@@ -365,10 +370,29 @@ static void oacp_ind_handler(struct bt_conn *conn,
-			net_buf_simple_pull_u8(&net_buf);
-		enum bt_gatt_ots_oacp_res_code  result_code =
-			net_buf_simple_pull_u8(&net_buf);
+		if (net_buf.len >= (sizeof(uint8_t) * 2)) {
```
```suggestion
		if (net_buf.len >= (sizeof(req_opcode) + sizeof(result_code))) {
```

pirun replied:
done

alwa-nordic commented on include/zephyr/bluetooth/services/ots.h:
```diff
@@ -725,6 +726,24 @@ struct bt_ots_cb {
+	 *
+	 *  @return 0 to accept, or any negative value to reject.
+	 */
+	int (*obj_cal_checksum)(struct bt_ots *ots, struct bt_conn *conn, uint64_t id,
```
Can't we just use the `obj_read` method?

pirun replied:
Considering obj_read might fragment data. like what we do in [sample ](https://github.com/zephyrproject-rtos/zephyr/blob/main/samples/bluetooth/peripheral_ots/src/main.c#L152).
I think it's better to have a dedicate callback for checksum,

jori-nordic commented on subsys/bluetooth/services/ots/ots_client.c:
```diff
@@ -1263,6 +1287,49 @@ static int oacp_write(struct bt_conn *conn, struct bt_otc_internal_instance_t *i
+{
+	int err;
+
+	LOG_DBG("oacp_checksum");
```
aren't you getting a checkpatch failure from this?

pirun replied:
No, would checkpatch suppress by git reuse-message?

pirun replied:
Removed

jori-nordic replied:
the standard form is just `LOG_DBG("")` since `LOG_DBG` will include the current function name. 
Checkpatch had a check that failed if the function name (here "oacp_checksum") was found in the log string. Maybe they temporarily disabled it, I remember there were a bunch of false-positives.

jori-nordic commented on subsys/bluetooth/services/ots/ots_client.c:
```diff
@@ -1263,6 +1287,49 @@ static int oacp_write(struct bt_conn *conn, struct bt_otc_internal_instance_t *i
+		return -EINVAL;
+	} else if (inst->busy) {
+		return -EBUSY;
+	} else if (cur_inst) {
```
you could similarly add debug prints to differentiate those two `-EBUSY` retcodes.

pirun replied:
Add two LOG_DBG to tell which busy statement is.

pirun committed:
Bluetooth: OTS - Add Calculate Checksum support

OTS add Calculate checksum feature support.
OTS client add object calculate checksum function.

Signed-off-by: Pirun Lee <pirun.lee@nordicsemi.no>

pirun committed:
sample: bluetooth: OTS sample add checksum

central_otc: read checksum after object data read.
peripheral_ots: add checksum calculate cb.

Signed-off-by: Pirun Lee <pirun.lee@nordicsemi.no>
